[aotd] HOP effect tokens wrapper above SubclassWrapper

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132755
* #132638
* __->__ #131672


Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.

IvanKobzarev committed:
[aotd] HOP effect tokens wrapper above SubclassWrapper

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -371,6 +373,80 @@ def add_hooks(module, config):
+        with torch.library._scoped_library("_mylib", "FRAGMENT") as lib:
+            lib.define("foo(Tensor x) -> Tensor")
+
+            def foo_impl(a):
+                torch.ops.aten._print("fwd")
+                return a.clone()
+
+            def foo_bwd(ctx, grad):
+                torch.ops.aten._print("bwd")
+                return grad.clone()
```
instead of printing, you could have these tick a counter on a global variable. Then, we can check that the global variables were mutated the correct amount in these tests.

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -371,6 +373,80 @@ def add_hooks(module, config):
+                    torch.tensor([4.0, 5.0, 6.0], requires_grad=True),
+                )
+
+            for ins_fn_req_grad in [ins_dense_req_grad, ins_sc_req_grad]:
+                ref_ins = ins_fn_req_grad()
+                ref_out = fn(*ref_ins)
+                ref_out.sum().backward()
+
+                compiled_fn = torch.compile(fn, backend="aot_eager")
+                ins = ins_fn_req_grad()
+                out = compiled_fn(*ins)
+                self.assertEqual(ref_out, out)
+                out.sum().backward()
+                self.assertEqual(ref_ins[1].grad, ins[1].grad)
+                self.assertEqual(ref_ins[0].grad, ins[0].grad)
```
Is it possible to do an self.assertExpectedInline test on this, with both the forward and backward graphs?

IvanKobzarev replied:
Added graph checks with assertExpectedInline

zou3519 reviewed (commented):
The approach LGTM, I want some more targeted tests (like one that checks the graphs)

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] HOP effect tokens wrapper above SubclassWrapper"



Original issue:
https://github.com/pytorch/pytorch/issues/129486

Before subclass_wrapper() got inputs containing additional effect tokens and failed as this did not match SubclassMeta indexes.

This happened as functionalization was responsible to add / remove those tokens. 

Functionalization can not be run above Subclasses, as args/outs are duplicated in case of mutations.


The main design thought is to  keep logic of EffectTokens, Subclasses, Functionalization to know as less as possible about each others transformations.

For that extracting EffectTokens manipulation to a separate wrapper, which will be processed above SubclassWrapper, while functionalization will happen below SubclassWrapper as before.

In that case subclass wrap/unwrap works without information of additional arguments.







[ghstack-poisoned]
